File: Presentation/InterviewExport.swift
import Cocoa
import UniformTypeIdentifiers

@available(macOS 14.0, *)
extension InterviewMasterDelegate {

    // MARK: - Export Interview

    @objc func exportInterview() {
        let exportableMessages = voiceMessages.filter { msg in
            switch msg.type {
            case .question, .answer, .followUp, .codingTask:
                return true
            case .userResponse, .status, .screenshot:
                return false
            }
        }

        guard !exportableMessages.isEmpty else {
            let alert = NSAlert()
            alert.messageText = "Nothing to Export"
            alert.informativeText = "Start an interview and have some Q&A before exporting."
            alert.alertStyle = .informational
            alert.addButton(withTitle: "OK")
            alert.runModal()
            return
        }

        let savePanel = NSSavePanel()
        savePanel.title = "Export Interview"
        savePanel.canCreateDirectories = true

        // Format selector accessory view
        let accessoryView = NSView(frame: NSRect(x: 0, y: 0, width: 250, height: 32))
        let label = NSTextField(labelWithString: "Format:")
        label.frame = NSRect(x: 0, y: 6, width: 55, height: 20)
        accessoryView.addSubview(label)

        let formatPopup = NSPopUpButton(frame: NSRect(x: 60, y: 2, width: 180, height: 28))
        formatPopup.addItems(withTitles: ["Markdown (.md)", "JSON (.json)"])
        formatPopup.target = nil
        formatPopup.tag = 100
        accessoryView.addSubview(formatPopup)

        savePanel.accessoryView = accessoryView

        // Set default filename
        let dateSuffix = formattedDateForFilename()
        savePanel.nameFieldStringValue = "interview_\(dateSuffix).md"
        savePanel.allowedContentTypes = [.plainText]

        // Update filename when format changes
        formatPopup.target = self
        formatPopup.action = #selector(exportFormatChanged(_:))
        // Store save panel reference in popup's identifier for the callback
        formatPopup.identifier = NSUserInterfaceItemIdentifier("exportFormatPopup")

        if savePanel.runModal() == .OK, let url = savePanel.url {
            let format: ExportFormat = formatPopup.indexOfSelectedItem == 1 ? .json : .markdown
            let useCase = ExportInterviewUseCase()
            let content = useCase.execute(messages: exportableMessages, format: format)

            do {
                try content.write(to: url, atomically: true, encoding: .utf8)
                NSWorkspace.shared.selectFile(url.path, inFileViewerRootedAtPath: "")
            } catch {
                let alert = NSAlert()
                alert.messageText = "Export Failed"
                alert.informativeText = error.localizedDescription
                alert.alertStyle = .critical
                alert.runModal()
            }
        }
    }

    @objc func exportFormatChanged(_ sender: NSPopUpButton) {
        // Update save panel filename extension based on format selection
        if let savePanel = sender.window?.sheetParent?.sheets.first as? NSSavePanel
            ?? NSApp.windows.compactMap({ $0 as? NSSavePanel }).first {
            let baseName = savePanel.nameFieldStringValue
                .replacingOccurrences(of: ".md", with: "")
                .replacingOccurrences(of: ".json", with: "")
            let ext = sender.indexOfSelectedItem == 1 ? ".json" : ".md"
            savePanel.nameFieldStringValue = baseName + ext
        }
    }

    func autoSaveSession() {
        let exportableMessages = voiceMessages.filter { msg in
            switch msg.type {
            case .question, .answer, .followUp, .codingTask:
                return true
            default:
                return false
            }
        }
        guard !exportableMessages.isEmpty else { return }

        let dir = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent("Documents/InterviewMaster/sessions")
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)

        let useCase = ExportInterviewUseCase()
        let json = useCase.execute(messages: exportableMessages, format: .json)
        let filename = "session_\(formattedDateForFilename()).json"
        let url = dir.appendingPathComponent(filename)

        try? json.write(to: url, atomically: true, encoding: .utf8)
        NSLog("📁 Auto-saved session to \(url.path)")

        // Analyze session and store in memory (async, non-blocking)
        if let client = anthropicClient {
            let analysis = SessionAnalysisUseCase()
            analysis.analyze(messages: exportableMessages, sourceFile: filename, anthropicClient: client)
        }
    }

    private func formattedDateForFilename() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd_HHmm"
        return formatter.string(from: Date())
    }

    private func cleanUserResponse(_ text: String) -> String {
        var cleaned = text

        while let first = cleaned.unicodeScalars.first, !first.isASCII || first.value < 32 {
            cleaned = String(cleaned.dropFirst())
        }

        let hallucationPrefixes = [
            "tabii,", "tabii", "tabibi", "merci beaucoup", "merci", "gracias",
            "thank you for watching", "thanks for watching", "subscribe",
            "reunited with", "accidental", "nexus,", "nexus"
        ]

        let lowerCleaned = cleaned.lowercased()
        for prefix in hallucationPrefixes {
            if lowerCleaned.hasPrefix(prefix) {
                cleaned = String(cleaned.dropFirst(prefix.count)).trimmingCharacters(in: .whitespaces)
                break
            }
        }

        if let firstChar = cleaned.first, firstChar.isLowercase {
            let words = cleaned.split(separator: " ", maxSplits: 10, omittingEmptySubsequences: true)
            for (index, word) in words.enumerated() {
                if let first = word.first, first.isUppercase && index > 0 {
                    let restOfSentence = words[index...].joined(separator: " ")
                    if restOfSentence.count > 10 {
                        cleaned = restOfSentence
                        break
                    }
                }
            }
        }

        return cleaned.trimmingCharacters(in: .whitespaces)
    }
}

File: Presentation/InterviewExport.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using System.Windows;
using Microsoft.Win32;
using InterviewMasterApp.Application.UseCases;
using InterviewMasterApp.Domain.Model;
using InterviewMasterApp.Presentation.Windows;

namespace InterviewMasterApp.Presentation
{
    /// <summary>
    /// Interview export functionality - save interview transcripts to Markdown or JSON.
    /// Ported from Swift: InterviewExport.swift
    /// Provides export dialog with format selection and auto-save capability.
    /// </summary>
    public class InterviewExport
    {
        private readonly List<InterviewMessage> _voiceMessages;

        public InterviewExport(List<InterviewMessage> voiceMessages)
        {
            _voiceMessages = voiceMessages ?? throw new ArgumentNullException(nameof(voiceMessages));
        }

        /// <summary>
        /// Export interview to file with user-selected format and location.
        /// </summary>
        public void ExportInterview()
        {
            var exportableMessages = _voiceMessages.Where(msg =>
                msg.Type == InterviewMessage.MessageType.Question ||
                msg.Type == InterviewMessage.MessageType.Answer ||
                msg.Type == InterviewMessage.MessageType.FollowUp ||
                msg.Type == InterviewMessage.MessageType.CodingTask
            ).ToList();

            if (exportableMessages.Count == 0)
            {
                MessageBox.Show(
                    "Start an interview and have some Q&A before exporting.",
                    "Nothing to Export",
                    MessageBoxButton.OK,
                    MessageBoxImage.Information);
                return;
            }

            var saveDialog = new SaveFileDialog
            {
                Title = "Export Interview",
                FileName = $"interview_{FormattedDateForFilename()}.md",
                Filter = "Markdown files (*.md)|*.md|JSON files (*.json)|*.json",
                FilterIndex = 1,
                AddExtension = true
            };

            if (saveDialog.ShowDialog() == true)
            {
                try
                {
                    // Determine format from file extension
                    var format = saveDialog.FileName.EndsWith(".json", StringComparison.OrdinalIgnoreCase)
                        ? ExportFormat.Json
                        : ExportFormat.Markdown;

                    var useCase = new ExportInterviewUseCase();
                    var content = useCase.Execute(exportableMessages, format);

                    File.WriteAllText(saveDialog.FileName, content);

                    // Open containing folder and select the file
                    System.Diagnostics.Process.Start("explorer.exe", $"/select,\"{saveDialog.FileName}\"");

                    StealthLogger.Shared.Log($"📁 Exported interview to {saveDialog.FileName}");
                }
                catch (Exception ex)
                {
                    MessageBox.Show(
                        ex.Message,
                        "Export Failed",
                        MessageBoxButton.OK,
                        MessageBoxImage.Error);

                    StealthLogger.Shared.Log($"❌ Export failed: {ex.Message}");
                }
            }
        }

        /// <summary>
        /// Auto-save session to Documents folder in JSON format.
        /// Called periodically during interview or when significant events occur.
        /// </summary>
        public void AutoSaveSession()
        {
            var exportableMessages = _voiceMessages.Where(msg =>
                msg.Type == InterviewMessage.MessageType.Question ||
                msg.Type == InterviewMessage.MessageType.Answer ||
                msg.Type == InterviewMessage.MessageType.FollowUp ||
                msg.Type == InterviewMessage.MessageType.CodingTask
            ).ToList();

            if (exportableMessages.Count == 0)
                return;

            try
            {
                // Create auto-save directory in Documents
                var documentsPath = Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments);
                var sessionDir = Path.Combine(documentsPath, "InterviewMaster", "sessions");
                Directory.CreateDirectory(sessionDir);

                var useCase = new ExportInterviewUseCase();
                var json = useCase.Execute(exportableMessages, ExportFormat.Json);

                var filename = $"session_{FormattedDateForFilename()}.json";
                var filePath = Path.Combine(sessionDir, filename);

                File.WriteAllText(filePath, json);

                StealthLogger.Shared.Log($"📁 Auto-saved session to {filePath}");
            }
            catch (Exception ex)
            {
                StealthLogger.Shared.Log($"⚠️ Auto-save failed: {ex.Message}");
                // Silent fail for auto-save - don't interrupt user
            }
        }

        /// <summary>
        /// Clean user response text from common ASR hallucinations and noise.
        /// Removes garbage prefixes, non-ASCII characters, and detects sentence boundaries.
        /// </summary>
        public static string CleanUserResponse(string text)
        {
            if (string.IsNullOrWhiteSpace(text))
                return string.Empty;

            var cleaned = text;

            // Remove leading non-ASCII or control characters
            while (cleaned.Length > 0)
            {
                var firstChar = cleaned[0];
                if (firstChar >= 32 && firstChar < 127)
                    break;
                cleaned = cleaned.Substring(1);
            }

            // Remove common ASR hallucination prefixes
            var hallucinationPrefixes = new[]
            {
                "tabii,", "tabii", "tabibi", "merci beaucoup", "merci", "gracias",
                "thank you for watching", "thanks for watching", "subscribe",
                "reunited with", "accidental", "nexus,", "nexus"
            };

            var lowerCleaned = cleaned.ToLowerInvariant();
            foreach (var prefix in hallucinationPrefixes)
            {
                if (lowerCleaned.StartsWith(prefix))
                {
                    cleaned = cleaned.Substring(prefix.Length).TrimStart();
                    break;
                }
            }

            // If first character is lowercase, try to find the actual sentence start
            if (cleaned.Length > 0 && char.IsLower(cleaned[0]))
            {
                var words = cleaned.Split(new[] { ' ' }, 11, StringSplitOptions.RemoveEmptyEntries);
                for (int i = 0; i < words.Length; i++)
                {
                    if (words[i].Length > 0 && char.IsUpper(words[i][0]) && i > 0)
                    {
                        var restOfSentence = string.Join(" ", words.Skip(i));
                        if (restOfSentence.Length > 10)
                        {
                            cleaned = restOfSentence;
                            break;
                        }
                    }
                }
            }

            return cleaned.Trim();
        }

        /// <summary>
        /// Format current date/time for filename (yyyy-MM-dd_HHmm).
        /// </summary>
        private string FormattedDateForFilename()
        {
            return DateTime.Now.ToString("yyyy-MM-dd_HHmm", CultureInfo.InvariantCulture);
        }
    }
}

